code/SConstruct

```diff
@@ -1,139 +1,205 @@
-              '\033[32mChecking dependencies\033[0m \\\r',
-              '\033[32mChecking dependencies\033[0m |\r',
-              '\033[32mChecking dependencies\033[0m /\r'], interval=5)
+env = Environment(
+    tools = ['default', 'cxxtest'],
+    shared = flag_shared,
+    CXX = opt_cxx,
+    SHCXX = opt_cxx, # must have both of these, along with LINK and SHLINK
+    CPPPATH = opt_includes,
+    LIBPATH = opt_libdirs,
+)
+Export('env')
 
-# command line help
-Help("""
-Targets:
-sneezy:         build the main binary (output: sneezy.2).
-lowtools:       build the low tools (output: lowtools/).
-objs/*/*.os:    build a specific file (eg: objs/game/game_hilo.os).
-tags:           rebuild the emacs tag file. 
-check:          build and run unit tests
-'-c'            to clean objects and executables.
-'--no-progress' to mute the "Checking dependencies" output.
-""")
+
+## detect compiler and version
+gcc_version = None
+clang_version = None
+if opt_cxx.startswith('g++'):
+    # dumpversion has more reliable formatting than --version
+    gcc_version = subprocess.check_output([opt_cxx, '-dumpversion'])
+    gcc_version = tuple(map(int, gcc_version.split('.')[:2]))
+elif opt_cxx.startswith('clang'):
+    # clang dumpversion always reports 4.2.1...
+    clang_version = subprocess.check_output([opt_cxx, '--version'])
+    clang_version = clang_version.partition('clang version ')[2]
+    clang_version = tuple(map(int, clang_version.split()[0].split('.')[:2]))
+else:
+    print >>sys.stderr, 'Unrecognized compiler'
+    raise SystemExit
+
+
+## debug: runtime bug detection with high overhead
+if flag_debug:
+    # undefined behavior sanitizer
+    env.Append(CXXFLAGS='-fsanitize=undefined')
+    env.Append(extra_libs=['ubsan'])
+    # AddressSanitizer
+    env.Append(extra_libs=['asan'])
+    env.Append(CXXFLAGS=['-fsanitize=address', '-fno-omit-frame-pointer',
+                         '-fno-common'])
+    if clang_version:
+        env.Append(CXXFLAGS='-fsanitize=integer')
+
+
+## harden: security flags, slows runtime a bit
+if flag_harden:
+    if flag_debug:
+        print >>sys.stderr, 'WARNING: not recommended to enable both debug and harden'
+    if gcc_version:
+        # at startup, turn some elf sections RO, and prebind dynamic bindings
+        env.Append(CXXFLAGS=['-Wl,-z,now', '-Wl,-z,relro'])
+        # buffer overflow checker, only works for -O1 or greater
+        env.Append(CXXFLAGS='-D_FORTIFY_SOURCE=2')
+    if not flag_shared:
+        # position-independent executable (shared mode already has -fPIC)
+        env.Append(CXXFLAGS='-fPIE', LINKFLAGS=['-fPIE', '-pie'])
+    # buffer overflow checker, only works for -O1 or greater
+    env.Append(CXXFLAGS='-D_FORTIFY_SOURCE=2')
+    # stack overflow checker
+env.Append(CXXFLAGS='-fstack-protector-strong')
+
+
+## optimize: compiles slow, bad for debugging, good for CPU
+opt_level = '-O1'
+if flag_optimize:
+    opt_level = '-O3'
```

Comment:
I think the biggest danger from -o3 is when people rely on "undefined behavior" in corner cases in the C spec, like signed integer overflow, where the compiler is free to say turn all your ints to 0xdeadbeef.